File: backend-nestjs/src/auth/auth.service.ts
import {
  BadRequestException,
  ConflictException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { RegisterDto } from './dto/register.dto';
import * as bcrypt from 'bcrypt';
import { JwtService } from '@nestjs/jwt';
import { LoginDto } from './dto/login.dto';
import { Verify2faDto } from './dto/verify-2fa.dto';
import { OTP } from 'otplib';
import { ForgotPasswordDto } from './dto/forgot-password.dto';
import { createHash, randomBytes } from 'crypto';
import { ResetPasswordDto } from './dto/reset-password.dto';
import type { PasswordReset } from '../../generated/prisma/client';
import { OAuth2Client } from 'google-auth-library';
import type { LoginTicket } from 'google-auth-library';
import { GoogleLoginDto } from './dto/google-login.dto';

@Injectable()
export class AuthService {
  constructor(
    private prisma: PrismaService,
    private jwtService: JwtService,
  ) {}

  private otp = new OTP();
  private googleClient = new OAuth2Client(process.env.GOOGLE_CLIENT_ID);

  async register(dto: RegisterDto) {
    if (dto.password !== dto.confirmPassword) {
      throw new BadRequestException('As duas senhas precisam ser iguais.');
    }

    const existingUser = await this.prisma.user.findUnique({
      where: { email: dto.email },
    });

    if (existingUser) {
      throw new ConflictException('Este email já está cadastrado.');
    }

    const hashedPassword = await bcrypt.hash(dto.password, 10);

    const usuario = await this.prisma.user.create({
      data: { name: dto.name, email: dto.email, password: hashedPassword },
      select: {
        id: true,
        name: true,
        email: true,
        role: true,
        twoFactorEnabled: true,
        createdAt: true,
        updatedAt: true,
      },
    });

    return usuario;
  }

  async login(dto: LoginDto) {
    const existingUser = await this.prisma.user.findUnique({
      where: { email: dto.email },
    });

    if (!existingUser || !existingUser.password) {
      throw new UnauthorizedException('Credenciais inválidas');
    }

    const comparePassword = await bcrypt.compare(
      dto.password,
      existingUser.password,
    );

    if (!comparePassword) {
      throw new UnauthorizedException('Credenciais inválidas');
    }

    if (existingUser.twoFactorEnabled) {
      const tempToken = this.jwtService.sign(
        { sub: existingUser.id, stage: '2fa-pending' },
        { expiresIn: '5m' },
      );

      return {
        requiresTwoFactor: true,
        tempToken,
      };
    }

    const accessToken = this.jwtService.sign({
      sub: existingUser.id,
      email: existingUser.email,
      role: existingUser.role,
    });

    const refreshToken = randomBytes(40).toString('hex');
    const refreshTokenHash = createHash('sha256')
      .update(refreshToken)
      .digest('hex');

    await this.prisma.refreshToken.create({
      data: {
        tokenHash: refreshTokenHash,
        userId: existingUser.id,
        expiresAt: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
      },
    });

    return { access_token: accessToken, refresh_token: refreshToken };
  }

  async verifyTwoFactor(dto: Verify2faDto) {
    let payload: { sub: string; stage?: string };

    try {
      payload = this.jwtService.verify(dto.tempToken);
    } catch {
      throw new UnauthorizedException('Token temporário inválido ou expirado');
    }

    if (payload.stage !== '2fa-pending') {
      throw new UnauthorizedException('Token inválido para está operação.');
    }

    const usuario = await this.prisma.user.findUnique({
      where: { id: payload.sub },
    });

    if (!usuario || !usuario.twoFactorSecret) {
      throw new UnauthorizedException(
        'Usuário inválido ou 2FA não configurado',
      );
    }

    const result = await this.otp.verify({
      secret: usuario.twoFactorSecret,
      token: dto.code,
    });

    if (!result.valid) {
      throw new UnauthorizedException('Código de autenticação inválido');
    }

    const token = this.jwtService.sign({
      sub: usuario.id,
      email: usuario.email,
      role: usuario.role,
    });

    return { access_token: token };
  }

  async forgotPassword(dto: ForgotPasswordDto) {
    const usuario = await this.prisma.user.findUnique({
      where: { email: dto.email },
    });

    if (usuario) {
      const token = randomBytes(32).toString('hex');
      const tokenHash = await bcrypt.hash(token, 10);
      const expiresAt = new Date(Date.now() + 15 * 60 * 1000);

      await this.prisma.passwordReset.create({
        data: {
          email: dto.email,
          tokenHash,
          expiresAt,
        },
      });

      console.log(
        `Link de recuperação (simulado): http://localhost:4200/reset-password?token=${token}`,
      );
    }

    return {
      message: 'Se o e-mail existir, um link de recuperação foi enviado.',
    };
  }

  async resetPassword(dto: ResetPasswordDto) {
    const resets = await this.prisma.passwordReset.findMany({
      where: { expiresAt: { gt: new Date() } },
    });

    let resetEncontrado: PasswordReset | null = null;

    for (const reset of resets) {
      const bate = await bcrypt.compare(dto.token, reset.tokenHash);
      if (bate) {
        resetEncontrado = reset;
        break;
      }
    }

    if (!resetEncontrado) {
      throw new BadRequestException('Token inválido ou expirado');
    }

    const hashedPassword = await bcrypt.hash(dto.newPassword, 10);

    await this.prisma.user.update({
      where: { email: resetEncontrado.email },
      data: { password: hashedPassword },
    });

    await this.prisma.passwordReset.delete({
      where: { id: resetEncontrado.id },
    });

    return { message: 'Senha atualizada com sucesso.' };
  }

  async loginWithGoogle(dto: GoogleLoginDto) {
    let ticket: LoginTicket;

    try {
      ticket = await this.googleClient.verifyIdToken({
        idToken: dto.idToken,
        audience: process.env.GOOGLE_CLIENT_ID,
      });
    } catch {
      throw new UnauthorizedException('Token do Google inválido');
    }

    const payload = ticket.getPayload();

    if (!payload || !payload.email) {
      throw new UnauthorizedException(
        'Não foi possível obter o e-mail da conta Google',
      );
    }

    const usuario = await this.prisma.user.upsert({
      where: { email: payload.email },
      update: {},
      create: {
        name: payload.name ?? 'Usuário Google',
        email: payload.email,
        password: null,
      },
    });

    const token = this.jwtService.sign({
      sub: usuario.id,
      email: usuario.email,
      role: usuario.role,
    });

    return { access_token: token };
  }

  async refresh(refreshTokenRecebido: string) {
    const tokenHash = createHash('sha256')
      .update(refreshTokenRecebido)
      .digest('hex');

    const tokenValido = await this.prisma.refreshToken.findFirst({
      where: { tokenHash, expiresAt: { gt: new Date() } },
    });

    if (!tokenValido) {
      throw new UnauthorizedException('Refresh token inválido');
    }

    const usuario = await this.prisma.user.findUnique({
      where: { id: tokenValido.userId },
    });

    if (!usuario) {
      throw new UnauthorizedException('Usuário não encontrado');
    }

    await this.prisma.refreshToken.delete({ where: { id: tokenValido.id } });

    const novoRefreshToken = randomBytes(40).toString('hex');
    const novoRefreshTokenHash = createHash('sha256')
      .update(novoRefreshToken)
      .digest('hex');
    await this.prisma.refreshToken.create({
      data: {
        tokenHash: novoRefreshTokenHash,
        userId: usuario.id,
        expiresAt: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
      },
    });

    const novoAccessToken = this.jwtService.sign({
      sub: usuario.id,
      email: usuario.email,
      role: usuario.role,
    });

    return { access_token: novoAccessToken, refresh_token: novoRefreshToken };
  }

  async logout(refreshTokenRecebido: string) {
    const tokenHash = createHash('sha256')
      .update(refreshTokenRecebido)
      .digest('hex');

    await this.prisma.refreshToken.deleteMany({
      where: { tokenHash },
    });

    return { message: 'Logout realizado com sucesso' };
  }
}

File: backend-springboot/backend-springboot/src/main/java/com/adryan/authbenchmark/backend_springboot/service/AuthService.java
package com.adryan.authbenchmark.backend_springboot.service;

import com.adryan.authbenchmark.backend_springboot.dto.LoginResponseDto;
import com.adryan.authbenchmark.backend_springboot.dto.TwoFactorPendingResponseDto;
import com.adryan.authbenchmark.backend_springboot.dto.TwoFactorVerifiedResponseDto;
import com.adryan.authbenchmark.backend_springboot.dto.UserResponseDto;
import com.adryan.authbenchmark.backend_springboot.exception.*;
import com.adryan.authbenchmark.backend_springboot.model.PasswordReset;
import com.adryan.authbenchmark.backend_springboot.model.RefreshToken;
import com.adryan.authbenchmark.backend_springboot.model.User;
import com.adryan.authbenchmark.backend_springboot.repository.PasswordResetRepository;
import com.adryan.authbenchmark.backend_springboot.repository.RefreshTokenRepository;
import com.adryan.authbenchmark.backend_springboot.repository.UserRepository;
import com.adryan.authbenchmark.backend_springboot.util.InputSanitizer;
import com.warrenstrange.googleauth.GoogleAuthenticator;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.time.LocalDateTime;
import java.util.HexFormat;
import java.util.List;
import java.util.UUID;

@Service
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final GoogleAuthenticator googleAuthenticator = new GoogleAuthenticator();
    private final PasswordResetRepository passwordResetRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final InputSanitizer inputSanitizer;

    public AuthService(UserRepository userRepository, PasswordEncoder passwordEncoder,  JwtService jwtService, PasswordResetRepository passwordResetRepository,  RefreshTokenRepository refreshTokenRepository, InputSanitizer inputSanitizer) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
        this.passwordResetRepository = passwordResetRepository;
        this.refreshTokenRepository = refreshTokenRepository;
        this.inputSanitizer = inputSanitizer;
    }

    public User register(String name, String email, String password, String confirmPassword){
        if(!password.equals(confirmPassword)){
            throw new PasswordMismatchException("As duas senhas precisam ser iguais.");
        }

        if(userRepository.findByEmail(email).isPresent()){
            throw new EmailAlreadyExistsException("Este email já está cadastrado.");
        }
        User user = new User();
        user.setName(inputSanitizer.sanitize(name));
        user.setEmail(email);
        user.setPassword(passwordEncoder.encode(password));

        return userRepository.save(user);
    }

    public Object login(String email, String password) {
        User user = userRepository.findByEmail(email).orElse(null);
        if (user == null || !passwordEncoder.matches(password, user.getPassword())) {
            throw new LoginFailedException("Credenciais inválidas");
        }

        if (user.isTwoFactorEnabled()) {
            String tempToken = jwtService.generateTempToken(user.getId());
            return new TwoFactorPendingResponseDto(true, tempToken);
        }

        String AccessToken = jwtService.generateToken(user.getId(), user.getEmail(), user.getRole().name());

        byte[] randomBytes = new byte[40];
        new SecureRandom().nextBytes(randomBytes);
        String refreshToken = HexFormat.of().formatHex(randomBytes);
        String refreshTokenHash = hashToken(refreshToken);

        RefreshToken refreshTokenEntity = new RefreshToken();
        refreshTokenEntity.setTokenHash(refreshTokenHash);
        refreshTokenEntity.setUserId(user.getId());
        refreshTokenEntity.setExpiresAt(LocalDateTime.now().plusDays(7));
        refreshTokenRepository.save(refreshTokenEntity);

        return new LoginResponseDto(AccessToken, refreshToken, new UserResponseDto(user));
    }

    public TwoFactorVerifiedResponseDto verifyTwoFactor(String tempToken, String code) {
        String stage;
        String userId;

        try {
            stage = jwtService.extractStage(tempToken);
            userId = jwtService.extractSubject(tempToken);
        } catch (Exception e) {
            throw new LoginFailedException("Token temporário inválido ou expirado");
        }

        if (!"2fa-pending".equals(stage)) {
            throw new LoginFailedException("Token inválido para esta operação");
        }

        User user = userRepository.findById(UUID.fromString(userId)).orElse(null);
        if (user == null || user.getTwoFactorSecret() == null) {
            throw new LoginFailedException("Usuário inválido ou 2FA não configurado");
        }

        boolean isValid = googleAuthenticator.authorize(user.getTwoFactorSecret(), Integer.parseInt(code));
        if (!isValid) {
            throw new LoginFailedException("Código de autenticação inválido");
        }

        String accessToken = jwtService.generateToken(user.getId(), user.getEmail(), user.getRole().name());
        return new TwoFactorVerifiedResponseDto(accessToken, new UserResponseDto(user));
    }

    public void forgotPassword(String email) {
        userRepository.findByEmail(email).ifPresent(user -> {
            byte[] randomBytes = new byte[32];
            new SecureRandom().nextBytes(randomBytes);
            String token = HexFormat.of().formatHex(randomBytes);
            String tokenHash = passwordEncoder.encode(token);

            PasswordReset passwordReset = new PasswordReset();
            passwordReset.setEmail(email);
            passwordReset.setTokenHash(tokenHash);
            passwordReset.setExpiresAt(LocalDateTime.now().plusMinutes(15));

            passwordResetRepository.save(passwordReset);

            System.out.println("Link de recuperação (simulado): http://localhost:4200/reset-password?token=" + token);
        });
    }

    public void resetPassword(String token, String newPassword) {
        List<PasswordReset> resets = passwordResetRepository.findByExpiresAtAfter(LocalDateTime.now());

        PasswordReset resetEncontrado = null;

        for(PasswordReset reset : resets){
            if(passwordEncoder.matches(token, reset.getTokenHash())){
                resetEncontrado = reset;
                break;
            }
        }

        if(resetEncontrado == null){
            throw new InvalidResetTokenException("Token inválido ou expirado");
        }

        User user = userRepository.findByEmail(resetEncontrado.getEmail())
                .orElseThrow(() -> new RuntimeException("Usuário não encontrado"));

        String hashedPassword = passwordEncoder.encode(newPassword);
        user.setPassword(hashedPassword);
        userRepository.save(user);

        passwordResetRepository.delete(resetEncontrado);
    }

    private String hashToken(String token) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(token.getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(hash);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("Erro ao gerar hash do token", e);
        }
    }

    public LoginResponseDto refresh(String refreshTokenRecebido) {
        String tokenHash = hashToken(refreshTokenRecebido);

        RefreshToken tokenValido = refreshTokenRepository
                .findByTokenHashAndExpiresAtAfter(tokenHash, LocalDateTime.now())
                .orElseThrow(() -> new InvalidRefreshTokenException("Refresh token inválido"));

        User user = userRepository.findById(tokenValido.getUserId())
                .orElseThrow(() -> new InvalidRefreshTokenException("Usuário não encontrado"));

        refreshTokenRepository.delete(tokenValido);

        byte[] randomBytes = new byte[40];
        new SecureRandom().nextBytes(randomBytes);
        String novoRefreshToken = HexFormat.of().formatHex(randomBytes);
        String novoRefreshTokenHash = hashToken(novoRefreshToken);

        RefreshToken novoRefreshTokenEntity = new RefreshToken();
        novoRefreshTokenEntity.setTokenHash(novoRefreshTokenHash);
        novoRefreshTokenEntity.setUserId(user.getId());
        novoRefreshTokenEntity.setExpiresAt(LocalDateTime.now().plusDays(7));
        refreshTokenRepository.save(novoRefreshTokenEntity);

        String novoAccessToken = jwtService.generateToken(user.getId(), user.getEmail(), user.getRole().name());

        return new LoginResponseDto(novoAccessToken, novoRefreshToken, new UserResponseDto(user));
    }

    @Transactional
    public void logout(String refreshTokenRecebido) {
        String tokenHash = hashToken(refreshTokenRecebido);
        refreshTokenRepository.deleteByTokenHash(tokenHash);
    }
}